sys/include/vfs.h

```diff
@@ -1013,6 +1007,35 @@ int vfs_normalize_path(char *buf, const char *path, size_t buflen);
  */
 const vfs_mount_t *vfs_iterate_mounts(const vfs_mount_t *cur);
 
+/**
+ * @brief Iterate through all mounted file systems by their root directories
+ *
+ * Unlike @ref vfs_iterate_mounts, this is thread safe, and allows thread safe
+ * access to the mount point's stats through @ref vfs_dstatvfs. If mounts or
+ * unmounts happen while iterating, this is guaranteed to report all file
+ * systems that stayed mounted, and may report any that are transiently
+ * mounted for up to as often as they are (re)mounted. Note that the volume
+ * being reported can not be unmounted as @p dir is an open directory.
+ *
+ * Zero-initialize @p dir to start. As long as @c true is returned, @p dir is a
+ * valid directory on which the user can call @ref vfs_readdir or @ref
+ * vfs_dstatvfs (or even peek at its `.mp` if they dare ignore the warning in
+ * @ref vfs_DIR).
+ *
+ * Users MUST NOT call @ref vfs_closedir if they intend to keep iterating, but
+ * MUST call it when aborting iteration.
```

Comment:
Would it make sense to require the user to always call ´vfs_closedir()` when they are done, no matter if they abort early or iterated over all mounts?

Reply:
Hm, it's a trade-off between complexity of the simple thing (just iterating over all, as `df` does) and loops with breaks (of which I wouldn't know an example, because if you want only one, you'd diropen or statvfs that right away). So I think that this way is the more practical one.

Reply:
IMHO 'you only need to use `vfs_closedir()` sometimes is inconsistent and will lead to it being forgotten even if needed.

`vfs_iterate_mount_dirs()` calls `vfs_closedir()` internally anyway, so why not remove it from there and always have it called by the user, so the behavior is always consistent. "When done iterating, call vfs_closedir()". Not  "When done iterating AND no more mount points are left, call vfs_closedir()"

Reply:
I'm just pushing back here because I have a hard time seeing why the loop would be aborted in the first place; can you imagine any situation why it would? Even when paginating mount points one'd need to go through all to get some consistency metric ... although OK, I can envision some patterns of pagination that *do* work with incomplete lists -- fine, will change it.

(After all it will make the error of someone forgetting to clean up more visible, as they can never unmount after the first iteration).

Reply:
I've been briefly pondering whether it makes sense to, instead of a boolean, return the dir argument. Then it'd behave more like any allocation function, for which "if it returns something, you must free it" (or pass it on, eg. to a realloc) is a more common pattern. Do you think that'd enhance usability? Binary-wise it'd be pretty much the same, provided the conditional branch tests for zeroness (which is the same for `==NULL` and `==false`).